channels_redis/core.py

```diff
@@ -310,66 +316,102 @@ def _setup_encryption(self, symmetric_encryption_keys):
             # Discard old channels based on group_expiry
             await connection.zremrangebyscore(key, min=0, max=int(time.time()) - self.group_expiry)
 
-            # Return current lot
             channel_names = [x.decode("utf8") for x in await connection.zrange(key, 0, -1)]
 
-        connection_to_channels, channel_to_message, channel_to_capacity, channel_to_key = \
+        channel_keys, connection_to_channel_keys, channel_keys_to_message, channel_keys_to_capacity = \
             self._map_channel_to_connection(channel_names, message)
 
-        for connection_index, channel_redis_keys in connection_to_channels.items():
+        for connection_index, channel_redis_keys in connection_to_channel_keys.items():
+
             # Create a LUA script specific for this connection.
             # Make sure to use the message specific to this channel, it is
             # stored in channel_to_message dict and contains the
             # __asgi_channel__ key.
+
             group_send_lua = """
-                for i=1,#KEYS do
-                    if redis.call('LLEN', KEYS[i]) < tonumber(ARGV[i + #KEYS]) then
-                        redis.call('RPUSH', KEYS[i], ARGV[i])
-                        redis.call('EXPIRE', KEYS[i], %d)
+                    for i=1,#KEYS do
+                        if redis.call('LLEN', KEYS[i]) < tonumber(ARGV[i + #KEYS]) then
+                            redis.call('RPUSH', KEYS[i], ARGV[i])
+                            redis.call('EXPIRE', KEYS[i], %d)
+                        end
                     end
-                end
-            """ % self.expiry
+                    """ % self.expiry
 
             # We need to filter the messages to keep those related to the connection
-            args = [channel_to_message[channel_name] for channel_name in channel_names
-                    if channel_to_key[channel_name] in channel_redis_keys]
+            args = [channel_keys_to_message[channel_key] for channel_key in channel_keys
+                    if channel_key in channel_redis_keys]
 
             # We need to send the capacity for each channel
-            args += [channel_to_capacity[channel_name] for channel_name in channel_names
-                    if channel_to_key[channel_name] in channel_redis_keys]
+            args += [channel_keys_to_capacity[channel_key] for channel_key in channel_keys
+                    if channel_key in channel_redis_keys]
 
+            # channel_keys does not contain a single redis key more than once
             async with self.connection(connection_index) as connection:
                 await connection.eval(group_send_lua, keys=channel_redis_keys, args=args)
```

Comment:
previously channel_redis_keys would contain duplicate keys as multiple channels may map to same redis key. hence the more number of pushes.